Mathlib/Data/Nat/Sqrt.lean

```diff
@@ -0,0 +1,370 @@
+#align nat.le_sqrt' Nat.le_sqrt'
+
+theorem sqrt_lt {m n : ℕ} : sqrt m < n ↔ m < n * n :=
+  lt_iff_lt_of_le_iff_le le_sqrt
+#align nat.sqrt_lt Nat.sqrt_lt
+
+theorem sqrt_lt' {m n : ℕ} : sqrt m < n ↔ m < n ^ 2 :=
+  lt_iff_lt_of_le_iff_le le_sqrt'
+#align nat.sqrt_lt' Nat.sqrt_lt'
+
+theorem sqrt_le_self (n : ℕ) : sqrt n ≤ n :=
+  le_trans (le_mul_self _) (sqrt_le n)
+#align nat.sqrt_le_self Nat.sqrt_le_self
+
+theorem sqrt_le_sqrt {m n : ℕ} (h : m ≤ n) : sqrt m ≤ sqrt n :=
+  le_sqrt.2 (le_trans (sqrt_le _) h)
+#align nat.sqrt_le_sqrt Nat.sqrt_le_sqrt
+
+@[simp]
+theorem sqrt_zero : sqrt 0 = 0 := rfl
+#align nat.sqrt_zero Nat.sqrt_zero
+
+theorem sqrt_eq_zero {n : ℕ} : sqrt n = 0 ↔ n = 0 :=
+  ⟨fun h =>
+    Nat.eq_zero_of_le_zero <| le_of_lt_succ <| (@sqrt_lt n 1).1 <| by rw [h]; exact by decide, by
+    rintro rfl
+    simp⟩
+#align nat.sqrt_eq_zero Nat.sqrt_eq_zero
+
+theorem eq_sqrt {n q} : q = sqrt n ↔ q * q ≤ n ∧ n < (q + 1) * (q + 1) :=
+  ⟨fun e => e.symm ▸ sqrt_IsSqrt n, fun ⟨h₁, h₂⟩ =>
+    le_antisymm (le_sqrt.2 h₁) (le_of_lt_succ <| sqrt_lt.2 h₂)⟩
+#align nat.eq_sqrt Nat.eq_sqrt
+
+theorem eq_sqrt' {n q} : q = sqrt n ↔ q ^ 2 ≤ n ∧ n < (q + 1) ^ 2 := by
+  simpa only [pow_two] using eq_sqrt
+#align nat.eq_sqrt' Nat.eq_sqrt'
+
+theorem le_three_of_sqrt_eq_one {n : ℕ} (h : sqrt n = 1) : n ≤ 3 :=
+  le_of_lt_succ <| (@sqrt_lt n 2).1 <| by rw [h]; exact by decide
+#align nat.le_three_of_sqrt_eq_one Nat.le_three_of_sqrt_eq_one
+
+theorem sqrt_lt_self {n : ℕ} (h : 1 < n) : sqrt n < n :=
+  sqrt_lt.2 <| by have := Nat.mul_lt_mul_of_pos_left h (lt_of_succ_lt h); rwa [mul_one] at this
+#align nat.sqrt_lt_self Nat.sqrt_lt_self
+
+theorem sqrt_pos {n : ℕ} : 0 < sqrt n ↔ 0 < n :=
+  le_sqrt
+#align nat.sqrt_pos Nat.sqrt_pos
+
+theorem sqrt_add_eq (n : ℕ) {a : ℕ} (h : a ≤ n + n) : sqrt (n * n + a) = n :=
+  le_antisymm
+    (le_of_lt_succ <|
+      sqrt_lt.2 <| by
+        rw [succ_mul, mul_succ, add_succ, add_assoc];
+          exact lt_succ_of_le (Nat.add_le_add_left h _))
+    (le_sqrt.2 <| Nat.le_add_right _ _)
+#align nat.sqrt_add_eq Nat.sqrt_add_eq
+
+theorem sqrt_add_eq' (n : ℕ) {a : ℕ} (h : a ≤ n + n) : sqrt (n ^ 2 + a) = n :=
+  (congr_arg (fun i => sqrt (i + a)) (sq n)).trans (sqrt_add_eq n h)
+#align nat.sqrt_add_eq' Nat.sqrt_add_eq'
+
+theorem sqrt_eq (n : ℕ) : sqrt (n * n) = n :=
+  sqrt_add_eq n (zero_le _)
+#align nat.sqrt_eq Nat.sqrt_eq
+
+theorem sqrt_eq' (n : ℕ) : sqrt (n ^ 2) = n :=
+  sqrt_add_eq' n (zero_le _)
+#align nat.sqrt_eq' Nat.sqrt_eq'
+
+@[simp]
+theorem sqrt_one : sqrt 1 = 1 :=
+  sqrt_eq 1
+#align nat.sqrt_one Nat.sqrt_one
+
+theorem sqrt_succ_le_succ_sqrt (n : ℕ) : sqrt n.succ ≤ n.sqrt.succ :=
+  le_of_lt_succ <|
+    sqrt_lt.2 <|
+      lt_succ_of_le <|
+        succ_le_succ <|
+          le_trans (sqrt_le_add n) <|
+            add_le_add_right
+              (by refine' add_le_add (Nat.mul_le_mul_right _ _) _ <;> exact Nat.le_add_right _ 2) _
```

Comment:
Is this not nicer, to prevent everything marching off to the right? (I have no experience with the style of mathlib4 and may be breaking conventions or something.)

```suggestion
  add_le_add_right
    (by refine' add_le_add (Nat.mul_le_mul_right _ _) _ <;> exact Nat.le_add_right _ 2) _
  |> le_trans (sqrt_le_add n)
  |> succ_le_succ
  |> lt_succ_of_le
  |> sqrt_lt.2
  |> le_of_lt_succ
```

Reply:
The more usual style is to just put several of these on the same line. Mathport is just using the default lean formatter and it is not that smart. I think it is better to avoid `|>`, it is rarely used so it will throw people off and the elaboration order is backwards (that is, `f` is elaborated first in `a |> f`) which is confusing.

Reply:
I have put several pipes on the same line in [my recent commit](https://github.com/leanprover-community/mathlib4/pull/1142/commits/1b8e9d8c155d73442765af096d998c87540ffdb4), as suggested by Mario Carneiro.